File: plugins/DungeonForge/src/main/java/nl/riddernix/dungeonforge/generation/GenerationStyle.java
package nl.riddernix.dungeonforge.generation;

import org.bukkit.Material;
import org.bukkit.block.data.BlockData;
import org.bukkit.configuration.file.FileConfiguration;

import java.util.EnumMap;
import java.util.Locale;
import java.util.Map;
import java.util.logging.Logger;

/** Materials for procedural rooms, fallback normal rooms, and temporary corridor platforms. */
public final class GenerationStyle {

    private final Map<DungeonLayout.RoomType, BlockData> roomMaterials;
    private final Map<String, BlockData> markerMaterials;
    private final BlockData corridorFloorMaterial;
    private final BlockData corridorLipMaterial;
    private final BlockData air = Material.AIR.createBlockData();

    private GenerationStyle(Map<DungeonLayout.RoomType, BlockData> roomMaterials, Map<String, BlockData> markerMaterials,
                            BlockData corridorFloorMaterial, BlockData corridorLipMaterial) {
        this.roomMaterials = Map.copyOf(roomMaterials);
        this.markerMaterials = Map.copyOf(markerMaterials);
        this.corridorFloorMaterial = corridorFloorMaterial;
        this.corridorLipMaterial = corridorLipMaterial;
    }

    public static GenerationStyle fromConfig(FileConfiguration config, Logger logger) {
        Map<DungeonLayout.RoomType, BlockData> rooms = new EnumMap<>(DungeonLayout.RoomType.class);
        rooms.put(DungeonLayout.RoomType.SPAWN,
                material(config, "generation.materials.entrance", Material.MOSS_BLOCK, logger));
        rooms.put(DungeonLayout.RoomType.NORMAL,
                material(config, "generation.materials.normal-room", Material.STONE, logger));
        rooms.put(DungeonLayout.RoomType.BRANCH,
                material(config, "generation.materials.branch", Material.STONE, logger));
        rooms.put(DungeonLayout.RoomType.BOSS,
                material(config, "generation.materials.boss", Material.REDSTONE_BLOCK, logger));
        Map<String, BlockData> markers = new java.util.HashMap<>();
        for (String category : new String[]{"swarm", "pack", "champion"}) {
            markers.put(category, material(config, "mobs.markers.materials." + category, Material.WHITE_WOOL, logger));
        }
        return new GenerationStyle(rooms, markers,
                material(config, "generation.corridor.floor-material", Material.STONE, logger),
                material(config, "generation.corridor.safety-lips.material", Material.STONE, logger));
    }

    public BlockData material(DungeonLayout.RoomType type) {
        return roomMaterials.get(type);
    }

    public BlockData corridorFloorMaterial() {
        return corridorFloorMaterial;
    }

    public BlockData corridorLipMaterial() {
        return corridorLipMaterial;
    }

    public BlockData air() {
        return air;
    }

    public BlockData marker(String category) {
        return markerMaterials.getOrDefault(category.toLowerCase(Locale.ROOT), Material.WHITE_WOOL.createBlockData());
    }

    private static BlockData material(FileConfiguration config, String path, Material fallback, Logger logger) {
        String raw = config.getString(path, fallback.name());
        Material material = raw == null ? null : Material.matchMaterial(raw.toUpperCase(Locale.ROOT));
        if (material == null || !material.isBlock()) {
            logger.warning("Unknown or unplaceable material at '" + path + "': " + raw
                    + ". Falling back to " + fallback + ".");
            material = fallback;
        }
        return material.createBlockData();
    }
}

File: plugins/DungeonPlugin/src/main/kotlin/nl/riddernix/dungeonplugin/generation/GenerationStyle.kt
package nl.riddernix.dungeonplugin.generation

import org.bukkit.Material
import org.bukkit.block.data.BlockData
import org.bukkit.configuration.file.FileConfiguration
import java.util.EnumMap
import java.util.Locale
import java.util.logging.Logger

/** Materials for procedural rooms, fallback normal rooms, and temporary corridor platforms. */
class GenerationStyle private constructor(
    roomMaterials: Map<DungeonLayout.RoomType, BlockData>,
    markerMaterials: Map<String, BlockData>,
    private val corridorFloorMaterial: BlockData,
    private val corridorLipMaterial: BlockData
) {

    private val roomMaterials: Map<DungeonLayout.RoomType, BlockData> = roomMaterials.toMap()
    private val markerMaterials: Map<String, BlockData> = markerMaterials.toMap()
    private val air: BlockData = Material.AIR.createBlockData()

    fun material(type: DungeonLayout.RoomType): BlockData = roomMaterials.getValue(type)

    fun corridorFloorMaterial(): BlockData = corridorFloorMaterial

    fun corridorLipMaterial(): BlockData = corridorLipMaterial

    fun air(): BlockData = air

    fun marker(category: String): BlockData =
        markerMaterials[category.lowercase(Locale.ROOT)] ?: Material.WHITE_WOOL.createBlockData()

    companion object {
        fun fromConfig(config: FileConfiguration, logger: Logger): GenerationStyle {
            val rooms = EnumMap<DungeonLayout.RoomType, BlockData>(DungeonLayout.RoomType::class.java)
            rooms[DungeonLayout.RoomType.SPAWN] =
                material(config, "generation.materials.entrance", Material.MOSS_BLOCK, logger)
            rooms[DungeonLayout.RoomType.NORMAL] =
                material(config, "generation.materials.normal-room", Material.STONE, logger)
            rooms[DungeonLayout.RoomType.BRANCH] =
                material(config, "generation.materials.branch", Material.STONE, logger)
            rooms[DungeonLayout.RoomType.BOSS] =
                material(config, "generation.materials.boss", Material.REDSTONE_BLOCK, logger)
            val markers = HashMap<String, BlockData>()
            for (category in arrayOf("swarm", "pack", "champion")) {
                markers[category] = material(config, "mobs.markers.materials.$category", Material.WHITE_WOOL, logger)
            }
            return GenerationStyle(rooms, markers,
                material(config, "generation.corridor.floor-material", Material.STONE, logger),
                material(config, "generation.corridor.safety-lips.material", Material.STONE, logger))
        }

        private fun material(config: FileConfiguration, path: String, fallback: Material, logger: Logger): BlockData {
            val raw = config.getString(path, fallback.name)
            var material = raw?.let { Material.matchMaterial(it.uppercase(Locale.ROOT)) }
            if (material == null || !material.isBlock) {
                logger.warning("Unknown or unplaceable material at '$path': $raw. Falling back to $fallback.")
                material = fallback
            }
            return material.createBlockData()
        }
    }
}
